catboost/catboost_pruning.py

```diff
@@ -0,0 +1,155 @@
+"""
+Optuna example that demonstrates a pruner for CatBoost.
+
+In this example, we optimize the validation accuracy of cancer detection using CatBoost.
+We optimize both the choice of booster model and their hyperparameters. Throughout
+training of models, a pruner observes intermediate results and stop unpromising trials.
+
+You can run this example as follows:
+    $ python catboost_pruning.py
+
+"""
+
+from typing import Any
+
+import optuna
+
+import catboost as cb
+from sklearn.datasets import load_breast_cancer
+from sklearn.metrics import roc_auc_score
+from sklearn.model_selection import train_test_split
+
+
+class CatBoostPruningCallback(object):
+    """Callback for catboost to prune unpromising trials.
+
+    You must call `check_pruned` after training manually unlike other pruning callbacks.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        metric:
+            An evaluation metric for pruning, e.g., ``Logloss`` and ``AUC``.
+            Please refer to
+            `CatBoost reference
+            <https://catboost.ai/docs/references/eval-metric__supported-metrics.html>`_
+            for further details.
+        valid_name:
+            The name of the target validation.
+            If you set only one ``eval_set``, ``validation`` is used.
+            If you set multiple datasets as ``eval_set``, the index number of ``eval_set`` must be
+            included in the valid_name, e.g., ``validation_0`` or ``validation_1``
+            when ``eval_set`` contains two datasets.
+    """
+
+    def __init__(
+        self, trial: optuna.trial.Trial, metric: str, valid_name: str = "validation"
+    ) -> None:
+        self._trial = trial
+        self._metric = metric
+        self._valid_name = valid_name
+        self._pruned = False
+        self._message = ""
+
+    def after_iteration(self, info: Any) -> bool:
+        """Run after each iteration.
+
+        Args:
+            info:
+                A simpleNamespace containing iteraion, validation_name, metric_name
```

Comment:
Typo and using inline code:
```suggestion
                A ``simpleNamespace`` containing iteration, ``validation_name``, ``metric_name``
```